apollo-router/src/query_planner/dual_query_planner.rs

```diff
@@ -0,0 +1,299 @@
+                } else {
+                    tracing::warn!("JS v.s. Rust query plan mismatch{operation_desc}");
+                    if let Some(formatted) = &js_plan.formatted_query_plan {
+                        tracing::debug!(
+                            "Diff of formatted plans:\n{}",
+                            render_diff(&diff::lines(formatted, &rust_plan.to_string()))
+                        );
+                    }
+                    tracing::trace!("JS query plan Debug: {js_root_node:#?}");
+                    tracing::trace!("Rust query plan Debug: {rust_root_node:#?}");
+                }
+            }
+        }
+
+        u64_counter!(
+            "apollo.router.operations.query_planner.both",
+            "Comparing JS v.s. Rust query plans",
+            1,
+            "generation.is_matched" = is_matched,
+            "generation.js_error" = self.js_result.is_err(),
+            "generation.rust_error" = rust_result.is_err()
+        );
+    }
+}
+
+// Specific comparison functions
+
+fn fetch_node_matches(this: &FetchNode, other: &FetchNode) -> bool {
+    let FetchNode {
+        service_name,
+        requires,
+        variable_usages,
+        operation,
+        operation_name,
+        operation_kind,
+        id,
+        input_rewrites,
+        output_rewrites,
+        context_rewrites,
+        schema_aware_hash: _, // ignored
+        authorization,
+    } = this;
+    *service_name == other.service_name
+        && *requires == other.requires
+        && *variable_usages == other.variable_usages
+        && *operation_name == other.operation_name
+        && *operation_kind == other.operation_kind
+        && *id == other.id
+        && *input_rewrites == other.input_rewrites
+        && *output_rewrites == other.output_rewrites
+        && *context_rewrites == other.context_rewrites
+        && *authorization == other.authorization
+        && operation_matches(operation, &other.operation)
+}
+
+fn subscription_primary_matches(this: &SubscriptionNode, other: &SubscriptionNode) -> bool {
+    let SubscriptionNode {
+        service_name,
+        variable_usages,
+        operation,
+        operation_name,
+        operation_kind,
+        input_rewrites,
+        output_rewrites,
+    } = this;
+    *service_name == other.service_name
+        && *variable_usages == other.variable_usages
+        && *operation_name == other.operation_name
+        && *operation_kind == other.operation_kind
+        && *input_rewrites == other.input_rewrites
+        && *output_rewrites == other.output_rewrites
+        && operation_matches(operation, &other.operation)
+}
+
+fn operation_matches(this: &SubgraphOperation, other: &SubgraphOperation) -> bool {
+    operation_without_whitespace(this) == operation_without_whitespace(other)
+}
+
+fn operation_without_whitespace(op: &SubgraphOperation) -> String {
+    op.as_serialized().replace([' ', '\n'], "")
+}
```

Comment:
i have a feeling we are going to be tweaking these as we go, but it seems like this is good to go from the current results in the corpus 🙏